### ISSUE TITLE
WARP prompt/mask token position

### ISSUE THREAD

**Issue by heyjoonkim** [YerevaNN/warp#1]
Hi,

Thank you for your work and for releasing the code! I just have a few questions regarding the ordering of the tokens.

From the script that you shared for training WARP, the prompt should be in the format of <code>[-10,-11,-12,-13,-14,null,-15,-16,-17,-18,-19,[MASK],-20,-21,-22,-23,-24,null,-25,-26,-27,-28,-29]</code>. If I understood correctly, the <code>hypothesis</code> and the <code>premise</code> should be placed in the <code>null</code> position and the other tokens should be considered as the prompt tokens. However, as I looked through the code I found that the input is given as <code>[[MASK], -10,-11,-12,-13,-14,-15,-16,-17,-18,-19,-20,-21,-22,-23,-24,-25,-26,-27,-28,-29, null, null]</code> with the <code>mask token</code> at the first position and the <code>hypothesis</code> and the <code>premise</code> at the end of the prompt. Did I understand the code correctly?  Could you clarify the positions of the prompt tokens?

Thanks again! 😄

**Comment by mahnerak** [comment]
@heyjoonkim Thanks for the question. I may add more detailed comments on how `reorder_optimized` flag works.

When the prompt template is `[-5,-6,null,-7,-8,[MASK],-9,null,-10,-11]` it actually means that `[MASK]` is inserted between the sentences. However, we tried to speed things up a little bit (`optimized_forward()` implementations of `ArpInjector`).

Right after tokenization, we permute all the [MASK] and prompt tokens to the beginning of the sentence ([MASK] always comes first), and their `position_id` is moved as well. Because we use Transformers when both token and its positional_ids are moved it makes zero difference.

As the result, in a single batch, all the `[MASK]` tokens stored in `inputs[:,0]` (fast pooling at masked position), all the prompt tokens are `inputs[:,1:1+num_prompt_tokens]` (fast application of trainable parameters) and the sentence tokens are `inputs[:,1+num_prompt_tokens:]` (application of the pretrained Embedding layer).

This is completely equivalent to what is described in the paper, but faster.

**Comment by heyjoonkim** [comment]
@mahnerak Thanks for the reply! It all makes sense now. Have a great day! 😄
